ports/zephyr/modzephyr.c

```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+    uint32_t block_num = mp_obj_get_int(args[1]);
+    off_t offset = block_num * self->block_size;
+    mp_buffer_info_t bufinfo;
+    int ret;
+
+    if (n_args >= 4) {
+        offset += mp_obj_get_int(args[3]);
+    }
+
+    mp_get_buffer_raise(args[2], &bufinfo, MP_BUFFER_WRITE);
+    ret = flash_area_read(self->area, offset, bufinfo.buf, bufinfo.len);
+    return MP_OBJ_NEW_SMALL_INT(ret);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(zephyr_flash_readblocks_obj, 3, 4, zephyr_flash_readblocks);
+
+STATIC mp_obj_t zephyr_flash_writeblocks(size_t n_args, const mp_obj_t *args) {
+    zephyr_flash_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    uint32_t block_num = mp_obj_get_int(args[1]);
+    off_t offset = block_num * self->block_size;
+    mp_buffer_info_t bufinfo;
+    int ret;
+
+    if (n_args == 3) {
+        if (flash_area_erase(self->area, offset, self->block_size)) {
+            mp_raise_ValueError("unable to erase block");
+        }
+    } else {
+        offset += mp_obj_get_int(args[3]);
+    }
+
+    mp_get_buffer_raise(args[2], &bufinfo, MP_BUFFER_READ);
+    ret = flash_area_write(self->area, offset, bufinfo.buf, bufinfo.len);
+    return MP_OBJ_NEW_SMALL_INT(ret);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(zephyr_flash_writeblocks_obj, 3, 4, zephyr_flash_writeblocks);
+
+STATIC mp_obj_t zephyr_flash_ioctl(mp_obj_t self_in, mp_obj_t cmd_in, mp_obj_t arg_in) {
+    zephyr_flash_obj_t *self = self_in;
+    mp_int_t cmd = mp_obj_get_int(cmd_in);
+    mp_int_t block_num = mp_obj_get_int(arg_in);
+
+    switch (cmd) {
+        case MP_BLOCKDEV_IOCTL_INIT:
+        case MP_BLOCKDEV_IOCTL_DEINIT:
+        case MP_BLOCKDEV_IOCTL_SYNC:
+            return MP_OBJ_NEW_SMALL_INT(0);
+
+        case MP_BLOCKDEV_IOCTL_BLOCK_COUNT:
+            return MP_OBJ_NEW_SMALL_INT(self->block_count);
+
+        case MP_BLOCKDEV_IOCTL_BLOCK_SIZE:
+            return MP_OBJ_NEW_SMALL_INT(self->block_size);
+
+        case MP_BLOCKDEV_IOCTL_BLOCK_ERASE:
+            if (flash_area_erase(self->area, block_num * self->block_size, self->block_size)) {
+                mp_raise_ValueError("unable to erase block");
+            }
+            return MP_OBJ_NEW_SMALL_INT(0);
+
+        default:
+            return MP_OBJ_NEW_SMALL_INT(-1);
+    }
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(zephyr_flash_ioctl_obj, zephyr_flash_ioctl);
+
+STATIC const mp_rom_map_elem_t zephyr_flash_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_readblocks), MP_ROM_PTR(&zephyr_flash_readblocks_obj) },
+    { MP_ROM_QSTR(MP_QSTR_writeblocks), MP_ROM_PTR(&zephyr_flash_writeblocks_obj) },
+    { MP_ROM_QSTR(MP_QSTR_ioctl), MP_ROM_PTR(&zephyr_flash_ioctl_obj) },
+#ifdef DT_FLASH_AREA_STORAGE_ID
+    { MP_ROM_QSTR(MP_QSTR_STORAGE), MP_ROM_INT(DT_FLASH_AREA_STORAGE_ID) },
+#endif
+};
+
+STATIC MP_DEFINE_CONST_DICT(zephyr_flash_locals_dict, zephyr_flash_locals_dict_table);
+
+STATIC const mp_obj_type_t zephyr_flash_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_Flash,
```

Comment:
As above with `Disk`, maybe call this class `FlashArea` to map to the Zephyr name?